### PR TITLE
gh-104050: Argument Clinic: Increase CConverter typing coverage

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-104050

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -2917,16 +2921,15 @@ def render(self, parameter: Parameter, data: CRenderData) -> None:
...
-        if not self.length:
-            return None
+        assert self.length is not None
         return self.parser_name + "_length"
 
     # Why is this one broken out separately?
     # For "positional-only" function parsing,
     # which generates a bunch of PyArg_ParseTuple calls.
-    def parse_argument(self, list):
+    def parse_argument(self, list: list[str]) -> None:
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: Okay but let's change the parameter name while we're here -- `list: list` looks ridiculous :P

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: I did that in an earlier version! 😆

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -2713,7 +2713,7 @@ class CConverter(metaclass=CConverterAutoRegister):
     # Or the magic value "unknown" if this value is a cannot be evaluated
     # at Argument-Clinic-preprocessing time (but is presumed to be valid
     # at runtime).
-    default: object = unspecified
+    default: object
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: Looks like removing this default here breaks things (see the failing tests), and I don't think removing _this_ one in particular buys us anything in terms of making it more mypy-friendly?

```suggestion
    default: object = unspecified
```

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: How about simply `args`?

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: True, there's no need to remove this one.

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (AlexWaygood) [reply Tools/clinic/clinic.py]: Works for me!

**Review by AlexWaygood — APPROVED** [review]
LGTM, providing the tests all pass with the default values removed!

**Comment by erlend-aasland** [comment]
Thanks for the review!

**Pull request merged**